File: TSL UHF ASCII 2 SDK DotNet v1.1/ASCII Protocol Inventory Sample VB/AsciiProtocolInventoryFx.ConvertedToVBNet/ViewModels/MainViewModel.vb
'-----------------------------------------------------------------------
' <copyright file="MainViewModel.cs" company="Technology Solutions UK Ltd"> 
'     Copyright (c) 2013 Technology Solutions UK Ltd. All rights reserved. 
' </copyright> 
' <author>Robin Stone</author>
'-----------------------------------------------------------------------
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports TechnologySolutions.Rfid.AsciiProtocol
Imports TechnologySolutions.Rfid.AsciiProtocol.Commands

''' <summary>
''' View model for main form
''' </summary>
Public Class MainViewModel
    Inherits PropertyChangedViewModel

    ''' <summary>
    ''' Initializes a new instance of the MainViewModel class
    ''' </summary>
    Public Sub New()
        Dim inventoryCommand As InventoryCommand
        Dim barcodeCommand As BarcodeCommand

        ' Create a display responder to capture and display all reader responses
        AddHandler Service.DisplayResponder.ReceivedLine, AddressOf Me.DisplayResponder_ReceivedLine

        ' setup an asynchronous responder for inventory
        inventoryCommand = New InventoryCommand()
        inventoryCommand.IsIndexedCommand = False
        inventoryCommand.IsLibraryCommand = False
        AddHandler inventoryCommand.TransponderReceived, AddressOf Me.AsynchronousTransponder_Received

        ' setup an asynchronous responder for barcodes
        barcodeCommand = New BarcodeCommand()
        barcodeCommand.IsIndexedCommand = False
        barcodeCommand.IsLibraryCommand = False
        AddHandler barcodeCommand.BarcodeReceived, AddressOf Me.AsynchronousBarcode_Received

        ' set up the responder chain
        Service.Reader.Commander.ClearResponders()
        Service.Reader.Commander.AddResponder(New LoggerResponder())
        Service.Reader.Commander.AddResponder(Service.DisplayResponder)
        Service.Reader.Commander.AddSynchronousResponder()
        Service.Reader.Commander.AddResponder(inventoryCommand.Responder)
        Service.Reader.Commander.AddResponder(barcodeCommand.Responder)
    End Sub

    Public Property Session() As QuerySession
        Get
            Return m_Session
        End Get

        Set(ByVal value As QuerySession)
            m_Session = value
        End Set
    End Property

    Private m_Session As QuerySession

    Public Property Target() As QueryTarget
        Get
            Return m_Target
        End Get

        Set(ByVal value As QueryTarget)
            m_Target = value
        End Set
    End Property

    Private m_Target As QueryTarget

    ''' <summary>
    ''' Raised to indicate transponder information to the user interface
    ''' </summary>
    Public Event TransponderMessageHandler As EventHandler(Of TextEventArgs)

    ''' <summary>
    ''' Raised to indicate transponder information to the user interface
    ''' </summary>
    Public Event BarcodeMessageHandler As EventHandler(Of TextEventArgs)

    ''' <summary>
    ''' Raised for each line of a response received
    ''' </summary>
    Public Event ResponseLine As EventHandler(Of TextEventArgs)

    ''' <summary>
    ''' Raised when a command finishes
    ''' </summary>
    Public Event CommandCompleted As EventHandler

    ''' <summary>
    ''' Execute the Inventory command
    ''' </summary>
    Public Sub ExecuteInventoryCommand(ByVal isCommandSynchronous As Boolean)
        Dim synchronousResponder As IAsciiCommandSynchronousResponder = Nothing

        Try
            ' Create a new Inventory command
            Dim inventory As New InventoryCommand()
            inventory.QuerySession = Me.Session
            inventory.QueryTarget = Me.Target

            If inventory IsNot Nothing Then
                ' 
                synchronousResponder = If(isCommandSynchronous, inventory.Responder, Nothing)
                AddHandler inventory.TransponderReceived, AddressOf Me.SynchronousTransponder_Received
            End If

            Try
                Service.Reader.Commander.ExecuteCommand(inventory, synchronousResponder)
            Finally
                If inventory IsNot Nothing Then
                    RemoveHandler inventory.TransponderReceived, AddressOf Me.SynchronousTransponder_Received
                End If

            End Try
        Catch ex As Exception
            Me.OnTransponderMessage(ex.Message)
        Finally
            Me.OnCommandCompleted()
        End Try
    End Sub

    Public Sub ExecuteBarcodeCommand(ByVal isCommandSynchronous As Boolean)
        Dim synchronousResponder As IAsciiCommandSynchronousResponder = Nothing

        Try
            ' Create a new Inventory command
            Dim bcCommand As New BarcodeCommand()

            If bcCommand IsNot Nothing Then
                ' 
                synchronousResponder = If(isCommandSynchronous, bcCommand.Responder, Nothing)
                AddHandler bcCommand.BarcodeReceived, AddressOf Me.SynchronousBarcode_Received
            End If

            Try
                Service.Reader.Commander.ExecuteCommand(bcCommand, synchronousResponder)
            Finally
                If bcCommand IsNot Nothing Then
                    RemoveHandler bcCommand.BarcodeReceived, AddressOf Me.SynchronousBarcode_Received
                End If
            End Try


            ' Report the error messages for unsuccessful synchronous commands
            If isCommandSynchronous AndAlso Not bcCommand.Response.IsSuccessful Then
                For Each message As String In bcCommand.Response.Messages
                    Me.OnBarcodeMessage(message)
                Next

            End If
        Catch ex As Exception
            Me.OnBarcodeMessage(ex.Message)
        Finally
            Me.OnCommandCompleted()
        End Try
    End Sub

    ''' <summary>
    ''' Raises the <see cref="CommandCompleted"/> event
    ''' </summary>
    Protected Overridable Sub OnCommandCompleted()
        RaiseEvent CommandCompleted(Me, EventArgs.Empty)
    End Sub

    ''' <summary>
    ''' Raises the <see cref="TransponderMessageHandler"/> event
    ''' </summary>
    ''' <param name="value">The message to display</param>
    Protected Overridable Sub OnTransponderMessage(ByVal value As String)
        RaiseEvent TransponderMessageHandler(Me, New TextEventArgs(value))
    End Sub

    ''' <summary>
    ''' Raises the <see cref="BarcodeMessageHandler"/> event
    ''' </summary>
    ''' <param name="value">The message to display</param>
    Protected Overridable Sub OnBarcodeMessage(ByVal value As String)
        RaiseEvent BarcodeMessageHandler(Me, New TextEventArgs(value))
    End Sub

    ''' <summary>
    ''' Raises the <see cref="ResponseLine"/> event
    ''' </summary>
    ''' <param name="value">The line to display</param>
    Protected Overridable Sub OnResponseLine(ByVal value As String)
        RaiseEvent ResponseLine(Me, New TextEventArgs(value))
    End Sub

    Private Sub AsynchronousTransponder_Received(ByVal sender As Object, ByVal e As TransponderDataEventArgs)
        Me.IssueTransponderMessage("[Async] ", e)
    End Sub

    Private Sub SynchronousTransponder_Received(ByVal sender As Object, ByVal e As TransponderDataEventArgs)
        Me.IssueTransponderMessage("[Sync] ", e)
    End Sub

    Private Sub IssueTransponderMessage(ByVal prefix As String, ByVal e As TransponderDataEventArgs)
        Dim message As String = String.Format(System.Globalization.CultureInfo.CurrentUICulture, "{0,-8} EPC: {1}", prefix, e.Transponder.Epc)
        Me.OnTransponderMessage(message)
    End Sub

    Private Sub AsynchronousBarcode_Received(ByVal sender As Object, ByVal e As BarcodeEventArgs)
        Me.IssueBarcodeMessage("[Async] ", e)
    End Sub

    Private Sub SynchronousBarcode_Received(ByVal sender As Object, ByVal e As BarcodeEventArgs)
        Me.IssueBarcodeMessage("[Sync] ", e)
    End Sub

    Private Sub IssueBarcodeMessage(ByVal prefix As String, ByVal e As BarcodeEventArgs)
        Dim message As String = String.Format(System.Globalization.CultureInfo.CurrentUICulture, "{0,-8} Barcode: {1}", prefix, e.Barcode)
        Me.OnBarcodeMessage(message)
    End Sub

    Private Sub DisplayResponder_ReceivedLine(ByVal sender As Object, ByVal e As AsciiLineEventArgs)
        Me.OnResponseLine(e.Line.FullLine)
    End Sub
End Class

File: TSL UHF ASCII 2 SDK DotNet v1.1/ASCII Protocol Inventory Sample/AsciiProtocolInventoryFx/ViewModels/MainViewModel.cs
//-----------------------------------------------------------------------
// <copyright file="MainViewModel.cs" company="Technology Solutions UK Ltd"> 
//     Copyright (c) 2013 Technology Solutions UK Ltd. All rights reserved. 
// </copyright> 
// <author>Robin Stone</author>
//-----------------------------------------------------------------------
namespace AsciiProtocolInventory.ViewModels
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    using TechnologySolutions.Rfid.AsciiProtocol;
    using TechnologySolutions.Rfid.AsciiProtocol.Commands;

    /// <summary>
    /// View model for main form
    /// </summary>
    public class MainViewModel
        : PropertyChangedViewModel
    {
        /// <summary>
        /// The ASCII commander
        /// </summary>
        private IAsciiCommandExecuting commander;

        /// <summary>
        /// The display settings
        /// </summary>
        private IDisplaySettings settings;

        /// <summary>
        /// Barcode command to perform a barcode scan synchronously
        /// </summary>
        private BarcodeCommand synchronousBarcodeCommand;

        /// <summary>
        /// Inventory command to perform an inventory scan asynchronously
        /// </summary>
        private InventoryCommand synchronousInventoryCommand;

        /// <summary>
        /// Initializes a new instance of the MainViewModel class
        /// </summary>
        /// <param name="displayResponder">Captures all responses from the reader and relays them to the user interface</param>
        /// <param name="commander">Used to setup the responder chain and execute ASCII commands</param>
        /// <param name="settings">The display settings</param>
        public MainViewModel(DisplayResponder displayResponder, IAsciiCommandExecuting commander, IDisplaySettings settings)
        {
            InventoryCommand inventoryResponder;
            BarcodeCommand barcodeResponder;

            if (displayResponder == null)
            {
                throw new ArgumentNullException("displayResponder");
            }

            if (commander == null)
            {
                throw new ArgumentNullException("commander");
            }

            if (settings == null)
            {
                throw new ArgumentNullException("settings");
            }

            this.settings = settings;

            // Create a display responder to capture and display all reader responses
            displayResponder.ReceivedLine += delegate(object sender, AsciiLineEventArgs e)
            {
                this.OnResponseLine(e.Line.FullLine);
            };
          
            // setup an asynchronous responder for inventory
            inventoryResponder = new InventoryCommand();
            inventoryResponder.TransponderReceived += this.AsynchronousTransponder_Received;            

            // setup an asynchronous responder for barcodes
            barcodeResponder = new BarcodeCommand();
            barcodeResponder.BarcodeReceived += this.AsynchronousBarcode_Received;

            // set up the responder chain
            this.commander = commander;
            this.commander.ClearResponders();
            this.commander.AddResponder(new LoggerResponder());
            this.commander.AddResponder(displayResponder);
            this.commander.AddSynchronousResponder();
            this.commander.AddResponder(inventoryResponder.Responder);
            this.commander.AddResponder(barcodeResponder.Responder);

            this.synchronousBarcodeCommand = new BarcodeCommand();
            this.synchronousBarcodeCommand.BarcodeReceived += this.SynchronousBarcode_Received;

            this.synchronousInventoryCommand = new InventoryCommand();
            this.synchronousInventoryCommand.TransponderReceived += this.SynchronousTransponder_Received;
        }        

        /// <summary>
        /// Raised to indicate transponder information to the user interface
        /// </summary>
        public event EventHandler<TextEventArgs> TransponderMessage;

        /// <summary>
        /// Raised to indicate transponder information to the user interface
        /// </summary>
        public event EventHandler<TextEventArgs> BarcodeMessage;

        /// <summary>
        /// Raised for each line of a response received
        /// </summary>
        public event EventHandler<TextEventArgs> ResponseLine;

        /// <summary>
        /// Gets or sets the qusery session to use for the inventory
        /// </summary>
        public QuerySession Session
        {
            get
            {
                return this.synchronousInventoryCommand.QuerySession ?? QuerySession.S0;
            }

            set
            {
                if (this.synchronousInventoryCommand.QuerySession != value)
                {
                    this.synchronousInventoryCommand.QuerySession = value;
                    this.OnPropertyChanged("Session");
                }
            }
        }

        /// <summary>
        /// Gets or sets the query targer to use for the inventory
        /// </summary>
        public QueryTarget Target 
        {
            get
            {
                return this.synchronousInventoryCommand.QueryTarget ?? QueryTarget.TargetA;
            }
            
            set
            {
                if (this.synchronousInventoryCommand.QueryTarget != value)
                {
                    this.synchronousInventoryCommand.QueryTarget = value;
                    this.OnPropertyChanged("Target");
                }
            }
        }

        /// <summary>
        /// Gets or sets a value indicating whether the responses panel should be visiable
        /// </summary>
        public bool IsProtocolResponseVisible
        {
            get
            {
                return this.settings.IsProtocolResponseVisible;
            }

            set
            {
                if (this.settings.IsProtocolResponseVisible != value)
                {
                    this.settings.IsProtocolResponseVisible = value;
                    this.OnPropertyChanged("IsProtocolResponseVisible");
                }
            }
        }

        /// <summary>
        /// Performs an inventory command synchronously
        /// </summary>
        public void ExecuteInventoryCommandSynchronously()
        {
            this.ExecuteInventoryCommand(true);
        }

        /// <summary>
        /// Performs an inventory command asynchronously
        /// </summary>
        public void ExecuteInventoryCommandAsynchronously()
        {
            this.ExecuteInventoryCommand(false);
        }

        /// <summary>
        /// Performs a barcode command synchronously
        /// </summary>
        public void ExecuteBarcodeCommandSynchronously()
        {
            this.ExecuteBarcodeCommand(true);
        }
                    
        /// <summary>
        /// Performs a barcode command asynchronously
        /// </summary>
        public void ExecuteBarcodeCommandAsynchronously()
        {
            this.ExecuteBarcodeCommand(false);
        }

        /// <summary>
        /// Cancels a currently executing barcode scan
        /// </summary>
        public void ExecuteAbortCommand()
        {
            try
            {
                this.commander.ExecuteCommand(new AbortCommand(), null);
            }
            catch (Exception ex)
            {
                this.OnBarcodeMessage(ex.Message);
            }
        }

        /// <summary>
        /// Raises the <see cref="TransponderMessageHandler"/> event
        /// </summary>
        /// <param name="value">The message to display</param>
        protected virtual void OnTransponderMessage(string value)
        {
            EventHandler<TextEventArgs> handler;

            handler = this.TransponderMessage;
            if (handler != null)
            {
                handler(this, new TextEventArgs(value));
            }
        }

        /// <summary>
        /// Raises the <see cref="BarcodeMessageHandler"/> event
        /// </summary>
        /// <param name="value">The message to display</param>
        protected virtual void OnBarcodeMessage(string value)
        {
            EventHandler<TextEventArgs> handler;

            handler = this.BarcodeMessage;
            if (handler != null)
            {
                handler(this, new TextEventArgs(value));
            }
        }

        /// <summary>
        /// Raises the <see cref="ResponseLine"/> event
        /// </summary>
        /// <param name="value">The line to display</param>
        protected virtual void OnResponseLine(string value)
        {
            EventHandler<TextEventArgs> handler;

            handler = this.ResponseLine;
            if (handler != null)
            {
                handler(this, new TextEventArgs(value));
            }
        }

        /// <summary>
        /// Execute the Inventory command
        /// </summary>
        /// <param name="isCommandSynchronous">True to execute synchronously, false to execute asychronously</param>
        /// <remarks>
        /// When executing synchronously the command's responder is passed to ExecuteCommand and the response is relayed back to the command
        /// When executing asynchronously no responder is passed to ExecuteCommand and the asynchronous responder in the responder chaain is
        /// left to capture the command's response
        /// </remarks>
        private void ExecuteInventoryCommand(bool isCommandSynchronous)
        {
            try
            {
                this.commander.ExecuteCommand(
                    this.synchronousInventoryCommand,
                    isCommandSynchronous ? this.synchronousInventoryCommand.Responder : null);
            }
            catch (Exception ex)
            {
                this.OnTransponderMessage(ex.Message);
            }
        }

        /// <summary>
        /// Execute the Barcode command
        /// </summary>
        /// <param name="isCommandSynchronous">True to execute synchronously, false to execute asychronously</param>
        /// <remarks>
        /// When executing synchronously the command's responder is passed to ExecuteCommand and the response is relayed back to the command
        /// When executing asynchronously no responder is passed to ExecuteCommand and the asynchronous responder in the responder chaain is
        /// left to capture the command's response
        /// </remarks>
        private void ExecuteBarcodeCommand(bool isCommandSynchronous)
        {
            try
            {
                this.commander.ExecuteCommand(
                        this.synchronousBarcodeCommand,
                        isCommandSynchronous ? this.synchronousBarcodeCommand.Responder : null);

                // Report the error messages for unsuccessful synchronous commands
                if (isCommandSynchronous && !this.synchronousBarcodeCommand.Response.IsSuccessful)
                {
                    foreach (string message in this.synchronousBarcodeCommand.Response.Messages)
                    {
                        this.OnBarcodeMessage(message);
                    }
                }
            }
            catch (Exception ex)
            {
                this.OnBarcodeMessage(ex.Message);
            }
        }

        /// <summary>
        /// Notifies the user interface of an asynchronous transponder response
        /// </summary>
        /// <param name="sender">The event source</param>
        /// <param name="e">Data provided for the event</param>
        private void AsynchronousTransponder_Received(object sender, TransponderDataEventArgs e)
        {
            this.IssueTransponderMessage("[Async] ", e);
        }

        /// <summary>
        /// Notifies the user interface of a synchronous transponder response
        /// </summary>
        /// <param name="sender">The event source</param>
        /// <param name="e">Data provided for the event</param>
        private void SynchronousTransponder_Received(object sender, TransponderDataEventArgs e)
        {
            this.IssueTransponderMessage("[Sync] ", e);
        }

        private void IssueTransponderMessage(string prefix, TransponderDataEventArgs e)
        {
            string message = string.Format(
                System.Globalization.CultureInfo.CurrentUICulture,
                 "{0,-8} EPC: {1}", 
                 prefix, 
                 e.Transponder.Epc);
            this.OnTransponderMessage(message);
        }

        /// <summary>
        /// Notifies the user interface of an asynchronous barcode response
        /// </summary>
        /// <param name="sender">The event source</param>
        /// <param name="e">Data provided for the event</param>
        private void AsynchronousBarcode_Received(object sender, BarcodeEventArgs e)
        {
            this.IssueBarcodeMessage("[Async] ", e);
        }

        /// <summary>
        /// Notifies the user interface of a synchronous barcode response
        /// </summary>
        /// <param name="sender">The event source</param>
        /// <param name="e">Data provided for the event</param>
        private void SynchronousBarcode_Received(object sender, BarcodeEventArgs e)
        {
            this.IssueBarcodeMessage("[Sync] ", e);
        }

        private void IssueBarcodeMessage(string prefix, BarcodeEventArgs e)
        {
            string message = string.Format(
                System.Globalization.CultureInfo.CurrentUICulture,
                "{0,-8} Barcode: {1}", 
                prefix, 
                e.Barcode);
            this.OnBarcodeMessage(message);
        }
    }
}
